source/core/slang-writer.cpp

```diff
@@ -71,6 +71,7 @@ ISlangUnknown* BaseWriter::getInterface(const Guid& guid)
 
 SLANG_NO_THROW char* SLANG_MCALL AppendBufferWriter::beginAppendBuffer(size_t maxNumChars)
 {
+    mutex.lock();
```

Comment:
Can we agree that the locking/unlocking stuff here maybe requires a comment or two given that:

* The `lock()` and `unlock()` calls aren't scoped or even in the same method, which makes sense given the invariants of the writer approach, but is maybe still surprising. (One might argue that the `beginAppend`/`endAppend` thing merits having some dedicated RAII type that captures the scoping and allows for an exception to flow out, perhaps aborting an append operation that was started but not completed...)
* There isn't a whole lot of thread-safety happening in the Slang codebase right now (for better or worse), and so it is probably worth making clear our intentions here: are we saying that all these basic utilities should start to have some minimal threading guarantees (and the question then becomes: what should those guarantees be?), or are we just doing it on an as-needed basis depending on which parts of our code need threading support?
* There's a final detail here that (I guess because the threading policy stuff here isn't stated), I'm left wondering if there are other operations here that should be mutex'd.

Reply:
Note that an alternative viewpoint here would be that the code that is creating an `AppendBufferWriter` and then handing it out to multiple threads should probably be the one that owns the mutex, and the threaded code itself should have a scoped lock on the mutex around where it does the append operations.

I'm guessing that isn't possible right now because the actual appends are deep in existing code that wasn't intended for threading.

Reply:
All of this rambling has me realizing that maybe the cleanest solution here from a design perspective is to have a wrapper `MutexControlledWriter` or similar, that wraps another arbitrary writer with this `lock()`/`unlock()` stuff.